File: nodegs/nodegs/BaseService.cpp
#include "stdafx.h"
#include "BaseService.h"
#include "Binary.h"
namespace OwLibSV
{
	_int64 m_downFlow = 0;

	_int64 m_upFlow = 0;

	MessageCallBack m_messageCallBack;

	map<int, BaseService*> m_services;

	WriteLogCallBack m_writeLogCallBack;

	CMessage::CMessage()
	{
		m_groupID = 0;
		m_serviceID = 0;
		m_functionID = 0;
		m_sessionID = 0;
		m_requestID = 0;
		m_socketID = 0;
		m_state = 0;
		m_compressType = 0;
		m_bodyLength = 0;
		m_body = 0;
	}

	CMessage::CMessage(int groupID, int serviceID, int functionID, int sessionID, int requestID, int socketID, int state, int compressType, int bodyLength, char *body)
	{
		m_groupID = groupID;
		m_serviceID = serviceID;
		m_functionID = functionID;
		m_sessionID = sessionID;
		m_requestID = requestID;
		m_socketID = socketID;
		m_state = state;
		m_compressType = compressType;
		m_bodyLength = bodyLength;
		m_body = body;
	}

	///////////////////////////////////////////////////////////////////////////////////////////////////////

	MessageListener::MessageListener()
	{
	}

	MessageListener::~MessageListener()
	{
		Clear();
	}

	void MessageListener::Add(ListenerMessageCallBack callBack, void *pInvoke)
	{
		m_callBacks.push_back(callBack);
		m_callBackInvokes.push_back(pInvoke);
	}

	void MessageListener::CallBack(CMessage *message)
	{
		int callBacksSize = (int)m_callBacks.size();
		for(int i = 0; i < callBacksSize; i++)
		{
			m_callBacks[i](message, m_callBackInvokes[i]);
		}
	}

	void MessageListener::Clear()
	{
		m_callBacks.clear();
		m_callBackInvokes.clear();
	}

	void MessageListener::Remove(ListenerMessageCallBack callBack)
	{
		void *pInvoke = 0;
		int pos = 0;
		vector<ListenerMessageCallBack>::iterator sIter = m_callBacks.begin();
		for(; sIter != m_callBacks.end(); ++sIter)
		{
			if(callBack == *sIter)
			{
				m_callBacks.erase(sIter);
				pInvoke = m_callBackInvokes[pos];
				break;
			}
			pos++;
		}
		vector<void*>::iterator sIter2 = m_callBackInvokes.begin();
		for(; sIter2 != m_callBackInvokes.end(); ++sIter2)
		{
			if(pInvoke == *sIter2)
			{
				m_callBackInvokes.erase(sIter2);
				break;
			}
		}
	}

	//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	SocketArray::SocketArray()
	{
	}

	SocketArray::~SocketArray()
	{
	}

	void SocketArray::AddSocket(int socketID)
	{
		vector<int>::iterator sIter = m_sockets.begin();
		for(; sIter != m_sockets.end(); ++sIter)
		{
			if(socketID == *sIter)
			{
				return;
			}
		}
		m_sockets.push_back(socketID);
	}

	void SocketArray::GetSocketList(vector<int> *socketList)
	{
		vector<int>::iterator sIter = m_sockets.begin();
		for(; sIter != m_sockets.end(); ++sIter)
		{
			socketList->push_back(*sIter);
		}
	}

	void SocketArray::RemoveSocket(int socketID)
	{
		vector<int>::iterator sIter = m_sockets.begin();
		for(; sIter != m_sockets.end(); ++sIter)
		{
			if(socketID == *sIter)
			{
				m_sockets.erase(sIter);
				return;
			}
		}
	}

	//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
	
	funcCloseServer p_funcCloseServer;
	funcRegisterCallBack p_funcRegisterCallBack;
	funcRegisterLog p_funcRegisterLog;
	funcSendByServer p_funcSendByServer;
	funcStartServer p_funcStartServer;

	int BaseService::CloseServer(int socketID)
	{
		return p_funcCloseServer(socketID);
	}

	void BaseService::RegisterCallBack(MessageCallBack callBack)
	{
		p_funcRegisterCallBack(callBack);
	}

	void BaseService::RegisterLog(WriteLogCallBack callBack)
	{
		p_funcRegisterLog(callBack);
	}

	int BaseService::SendByServer(int socketID, const char *str, int len)
	{
		return p_funcSendByServer(socketID, str, len);
	}

	int BaseService::StartServer(int type, int port)
	{
		return p_funcStartServer(type, port);
	}

	//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	BaseService::BaseService()
	{
		m_compressType = COMPRESSTYPE_GZIP;
		m_groupID = 0;
		m_serviceID = 0;
		m_sessionID = 0;
	}

	BaseService::~BaseService()
	{
		m_lock.Lock();
		map<int, MessageListener*>::iterator sIter = m_listeners.begin();
		for(; sIter != m_listeners.end(); ++sIter)
		{
			delete sIter->second;
		}
		m_listeners.clear();
		m_waitMessages.clear();
		m_lock.UnLock();
	}

	int BaseService::GetCompressType()
	{
		return 1;
	}

	void BaseService::SetCompressType(int compressType)
	{
		m_compressType = compressType;
	}

	_int64 BaseService::GetDownFlow()
	{
		return m_downFlow;
	}

	void BaseService::SetDownFlow(_int64 downFlow)
	{
		m_downFlow = downFlow;
	}

	int BaseService::GetGroupID()
	{
		return m_groupID;
	}

	void BaseService::SetGroupID(int groupID)
	{
		m_groupID = groupID;
	}

	int BaseService::GetServiceID()
	{
		return m_serviceID;
	}

	void BaseService::SetServiceID(int serviceID)
	{
		m_serviceID = serviceID;
	}

	int BaseService::GetSessionID()
	{
		return m_sessionID;
	}

	void BaseService::SetSessionID(int sessionID)
	{
		m_sessionID = sessionID;
	}

	_int64 BaseService::GetUpFlow()
	{
		return m_upFlow;
	}

	void BaseService::SetUpFlow(_int64 upFlow)
	{
		m_upFlow = upFlow;
	}

	/////////////////////////////////////////////////////////////////////////////////////////////////////////

	void BaseService::AddService(BaseService *service)
	{
		if (!m_messageCallBack)
		{
			string path = GetProgramDir() + "\\owsock_server.dll";
			HINSTANCE hdll = LoadLibraryA(path.c_str());
			p_funcCloseServer = (funcCloseServer)GetProcAddress(hdll, "CloseServer");
			p_funcRegisterCallBack = (funcRegisterCallBack)GetProcAddress(hdll, "RegisterServerCallBack");
			p_funcRegisterLog = (funcRegisterLog)GetProcAddress(hdll, "RegisterServerLog");
			p_funcSendByServer = (funcSendByServer)GetProcAddress(hdll, "SendByServer");
			p_funcStartServer = (funcStartServer)GetProcAddress(hdll, "StartServer");
			MessageCallBack callBack = &CallBack;
			m_messageCallBack = callBack;
			RegisterCallBack(callBack);
			WriteLogCallBack writeLogCallBack = &WriteServerLog;
			m_writeLogCallBack = writeLogCallBack;
			RegisterLog(m_writeLogCallBack);
		}
		m_services[service->GetServiceID()] = service;
	}

	void BaseService::CallBack(int socketID, int localSID, const char *str, int len)
	{
		m_downFlow += len;
		if (len > 4)
        {
			Binary binary;
			binary.Write(str, len);
			int head = binary.ReadInt();
			int groupID = (int)binary.ReadShort();
			int serviceID = (int)binary.ReadShort();
			BaseService *service = 0;
			if(m_services.find(serviceID) != m_services.end())
			{
				m_services[serviceID]->OnCallBack(&binary, socketID, localSID, len);
			}
		}
	}

	string BaseService::GetProgramDir()
	{
		char exeFullPath[MAX_PATH]; 
		string strPath = "";
		GetModuleFileNameA(0, exeFullPath, MAX_PATH);
		strPath = (string)exeFullPath; 
		int pos = (int)strPath.rfind('\\', strPath.length());
		return strPath.substr(0, pos);
	}


	int BaseService::GetRequestID()
	{
		return m_requestID ++;
	}

	void BaseService::GetServices(vector<BaseService*> *services)
	{
		map<int, BaseService*>::iterator sIter = m_services.begin();
		for(; sIter != m_services.end(); ++sIter)
		{
			services->push_back(sIter->second);
		}
	}

	int BaseService::KeepAlive(int socketID)
	{
		Binary binary;
		binary.Initialize(4, false);
		binary.WriteInt(4);
		char *str = (char*)binary.GetBytes();
		return SendByServer(socketID, str, 4);
	}

	void BaseService::OnCallBack(Binary *binary, int socketID, int localSID, int len)
	{
		int headSize = sizeof(int) * 4 + sizeof(short) * 3 + sizeof(byte) * 2;
		int functionID = (int)binary->ReadShort();
		int sessionID = binary->ReadInt();
		int requestID = binary->ReadInt();
		int state = (int)binary->ReadChar();
		int compressType = (int)binary->ReadChar();
		int uncBodyLength = binary->ReadInt();
		int bodyLength = len - headSize;
		char *body = new char[bodyLength];
		memset(body, '\0', bodyLength);
		binary->Read(body, bodyLength);
		if (compressType == COMPRESSTYPE_GZIP)
		{
			uLong cLen = 128;
			if(uncBodyLength > cLen)
			{
				cLen = uncBodyLength;
			}
			char *buffer = new char[cLen];
			memset(buffer, '\0', cLen);
			CStrA::GZDeCompress((Byte*)body, bodyLength, (Byte*)buffer, &cLen);
			delete[] body;
			body = buffer;
			uncBodyLength = cLen;
		}
		CMessage message(GetGroupID(), GetServiceID(), functionID, sessionID, requestID, socketID, state, compressType, uncBodyLength, body);
		OnReceive(&message);
		OnWaitMessageHandle(&message);
		if(message.m_body)
		{
			delete[] body;
		}
		body = 0;
	}

	void BaseService::OnClientClose(int socketID, int localSID)
	{
	}

	void BaseService::OnReceive(CMessage *message)
	{
		m_lock.Lock();
		m_compressTypes[message->m_socketID] = message->m_compressType;
		m_lock.UnLock();
	}

	void BaseService::OnWaitMessageHandle(CMessage *message)
	{
		if((int)m_waitMessages.size() > 0)
		{
			m_lock.Lock();
			map<int, CMessage*>::iterator sIter = m_waitMessages.find(message->m_requestID);
			if(sIter != m_waitMessages.end())
			{
				CMessage *waitMessage = sIter->second;
				waitMessage->Copy(message);
				char *newBody = new char[message->m_bodyLength];
				for (int i = 0; i < message->m_bodyLength; i++)
				{
					newBody[i] = message->m_body[i];
				}
				waitMessage->m_body = newBody;
			}
			m_lock.UnLock();
		}
	}

	void BaseService::RegisterWait(int requestID, CMessage *message)
	{
		m_lock.Lock();
		m_waitMessages[requestID] = message;
		m_lock.UnLock();
	}

	void BaseService::RegisterListener(int requestID, ListenerMessageCallBack callBack, void *pInvoke)
	{
		m_lock.Lock();
		MessageListener *listener = 0;
		map<int, MessageListener*>::iterator sIter = m_listeners.find(requestID);
		if(sIter != m_listeners.end())
		{
			listener = sIter->second;
		}
		else
		{
			listener = new MessageListener;
			m_listeners[requestID] = listener;
		}
		listener->Add(callBack, pInvoke);
		m_lock.UnLock();
	}

	int BaseService::Send(CMessage *message)
	{
		Binary binary;
		char *buffer = 0;
		const char *body = message->m_body;
		int bodyLength = message->m_bodyLength;
		int uncBodyLength = bodyLength;
		m_lock.Lock();
		map<int, int>::iterator sIter = m_compressTypes.find(message->m_socketID);
		if(sIter != m_compressTypes.end())
		{
			message->m_compressType = sIter->second;
		}
		m_lock.UnLock();
		if (message->m_compressType == COMPRESSTYPE_GZIP)
        {
			uLong cLen = bodyLength;
			buffer = new char[cLen];
			memset(buffer, '\0', cLen);
			CStrA::GZCompress((Byte*)body, bodyLength, (Byte*)buffer, &cLen);
			body = buffer;
			bodyLength = (int)cLen;
		}
		int len = sizeof(int) * 4 + bodyLength + sizeof(short) * 3 + sizeof(char) * 2;
		binary.Initialize(len, false);
		binary.WriteInt(len);
		binary.WriteShort((short)message->m_groupID);
		binary.WriteShort((short)message->m_serviceID);
		binary.WriteShort((short)message->m_functionID);
		binary.WriteInt(message->m_sessionID);
		binary.WriteInt(message->m_requestID);
		binary.WriteChar((char)message->m_state);
		binary.WriteChar((char)message->m_compressType);
		binary.WriteInt(uncBodyLength);
		binary.Write(body, bodyLength);
		char *str = (char*)binary.GetBytes();
		if(buffer)
		{
			delete[] buffer;
			buffer = 0;
		}
		int ret = SendByServer(message->m_socketID, str, len);
		m_upFlow += ret;
		return ret;
	}

	void BaseService::SendToListener(CMessage *message)
	{
		MessageListener *listener = 0;
		void *pInvoke = 0;
		m_lock.Lock();
		map<int, MessageListener*>::iterator sIter = m_listeners.find(message->m_requestID);
		if(sIter != m_listeners.end())
		{
			listener = sIter->second;
		}
		m_lock.UnLock();
		if(listener)
		{
			listener->CallBack(message);
		}
	}

	void BaseService::UnRegisterListener(int requestID)
	{
		m_lock.Lock();
		map<int, MessageListener*>::iterator sIter = m_listeners.find(requestID);
		if(sIter != m_listeners.end())
		{
			delete sIter->second;
			m_listeners.erase(sIter);
		}
		m_lock.UnLock();
	}

	void BaseService::UnRegisterListener(int requestID, ListenerMessageCallBack callBack)
	{
		m_lock.Lock();
		map<int, MessageListener*>::iterator sIter = m_listeners.find(requestID);
		if(sIter != m_listeners.end())
		{
			sIter->second->Remove(callBack);
		}
		m_lock.UnLock();
	}

	void BaseService::UnRegisterWait(int requestID)
	{
		m_lock.Lock();
		map<int, CMessage*>::iterator sIter = m_waitMessages.find(requestID);
		if(sIter != m_waitMessages.end())
		{
			m_waitMessages.erase(sIter);
		}
		m_lock.UnLock();
	}

	int BaseService::WaitMessage(int requestID, int timeout)
	{
		int state = 0;
		while(timeout > 0)
		{
			m_lock.Lock();
			map<int, CMessage*>::iterator sIter = m_waitMessages.find(requestID);
			if(sIter != m_waitMessages.end())
			{
				CMessage *message = sIter->second;
				if(message->m_bodyLength > 0)
				{
					state = 1;
					m_lock.UnLock();
					break;
				}
			}
			else
			{
				m_lock.UnLock();
				break;
			}
			m_lock.UnLock();
			timeout -= 10;
			Sleep(10);
		}
		UnRegisterWait(requestID);
		return state;
	}

	void BaseService::WriteServerLog(int socketID, int localSID, int state, const char *log)
	{
		if(state == 2)
		{
			map<int, BaseService*>::iterator sIter = m_services.begin();
			for(; sIter != m_services.end(); ++sIter)
			{
				BaseService *service = sIter->second;
				service->OnClientClose(socketID, localSID);
			}
		}
	}
}

File: nodegs_java/node.gs/src/nodegs/base/BaseService.java
//GAIA_LORDSETTLEMENT_JAVA_S1:LORD;
//****************************************************************************\
//*                                                                             *
//* BaseService.java -  BaseService functions, types, and definitions *
//*                                                                             *
//*               Version 1.00 ★                                                              	                        *
//*                                                                             *
//*               Copyright (c) 2017-2017, Server. All rights reserved.         *
//*               Created by tanfeng 2017/06/05.                                *
//*                                                                             *
//******************************************************************************

package nodegs.base;

import owchart.owlib.Sock.*;
//****************************************************************************\
//*                                                                             *
//* BaseService.cs -  Base service functions, types, and definitions            *
//*                                                                             *
//*               Version 1.00 ★★★★★                                       *
//*                                                                             *
//*               Copyright (c) 2016-2016, Server. All rights reserved.         *
//*               Created by Lord.                                              *
//*                                                                             *
//******************************************************************************

import java.io.*;
import java.util.*;

/**
 * 基础服务类
 * 
 */
public class BaseService {
	/**
	 * 创建服务
	 * 
	 */
	public BaseService() {
	}

	/**
	 * 析构函数
	 * 
	 */
	protected void finalize() throws Throwable {
		synchronized (m_listeners) {
			m_listeners.clear();
		}
		synchronized (m_waitMessages) {
			m_waitMessages.clear();
		}
	}

	/**
	 * 无压缩
	 * 
	 */
	public static int COMPRESSTYPE_NONE = 0;

	/**
	 * GZIP压缩
	 * 
	 */
	public static int COMPRESSTYPE_GZIP = 1;
	
	/**
	 * 压缩类型
	 */
	protected HashMap<Integer, Integer> m_compressTypes = new HashMap<Integer, Integer>();

	/**
	 * 监听者集合
	 * 
	 */
	private java.util.HashMap<Integer, MessageListener> m_listeners = new java.util.HashMap<Integer, MessageListener>();

	/**
	 * 请求ID
	 * 
	 */
	private static int m_requestID = 10000;

	/**
	 * 所有的服务
	 * 
	 */
	private static java.util.HashMap<Integer, BaseService> m_services = new java.util.HashMap<Integer, BaseService>();

	/**
	 * 等待消息队列
	 * 
	 */
	private java.util.HashMap<Integer, CMessage> m_waitMessages = new java.util.HashMap<Integer, CMessage>();

	private int m_compressType = COMPRESSTYPE_GZIP;

	/**
	 * 获取或设置压缩类型
	 * 
	 */
	public final int GetCompressType() {
		return m_compressType;
	}

	public final void SetCompressType(int value) {
		m_compressType = value;
	}

	protected static long m_downFlow;

	/**
	 * 获取或设置下载流量
	 * 
	 */
	public static long GetDownFlow() {
		return BaseService.m_downFlow;
	}

	public static void SetDownFlow(long value) {
		BaseService.m_downFlow = value;
	}

	private int m_groupID = 0;

	/**
	 * 获取或设置组ID
	 * 
	 */
	public final int GetGroupID() {
		return m_groupID;
	}

	public final void SetGroupID(int value) {
		m_groupID = value;
	}

	private boolean m_isDisposed = false;

	/**
	 * 获取对象是否已被销毁
	 * 
	 */
	public final boolean IsDisposed() {
		return m_isDisposed;
	}

	private int m_serviceID = 0;

	/**
	 * 获取或设置服务的ID
	 * 
	 */
	public final int GetServiceID() {
		return m_serviceID;
	}

	public final void SetServiceID(int value) {
		m_serviceID = value;
	}

	private int m_sessionID = 0;

	/**
	 * 获取或设置登录ID
	 * 
	 */
	public final int GetSessionID() {
		return m_sessionID;
	}

	public final void SetSessionID(int value) {
		m_sessionID = value;
	}

	private int m_socketID = 0;

	public final int GetSocketID() {
		return m_socketID;
	}

	public final void SetSocketID(int socketID) {
		m_socketID = socketID;
	}

	protected static long m_upFlow;

	/**
	 * 获取或设置上传流量
	 * 
	 */
	public static long GetUpFlow() {
		return m_upFlow;
	}

	public static void GetUpFlow(long value) {
		m_upFlow = value;
	}

	/**
	 * 关闭
	 * 
	 * @param socketID
	 *            连接ID
	 * @return 状态
	 */
	public static int CloseServer(int socketID) {
		return Servers.Close(socketID);
	}

	/**
	 * 发送消息
	 * 
	 * @param socketID
	 *            连接ID
	 * @param str
	 *            数据
	 * @param len
	 *            长度
	 * @return 状态
	 */
	public static int SendByServer(int socketID, int localSID, byte[] str, int len) {
		return Servers.Send(socketID, localSID, str, len);
	}

	/**
	 * 启动服务
	 * 
	 * @param port
	 *            端口号
	 * @return 状态
	 */
	public static int StartServer(int port) {
		Servers.SetListener(new ServerSocketListener());
		return Servers.Start(0, port);
	}

	/**
	 * 添加服务
	 * 
	 * @param service
	 *            服务
	 */
	public static void AddService(BaseService service) {
		m_services.put(service.GetServiceID(), service);
	}

	/**
	 * 回调函数
	 * 
	 * @param socketID
	 *            连接ID
	 * @param localSID
	 *            本地连接ID
	 * @param str
	 *            数据
	 * @param len
	 *            长度
	 */
	public static void CallBack(int socketID, int localSID, byte[] str, int len) {
		m_downFlow += len;
		try {
			if (len > 4) {
				Binary br = new Binary();
				br.Write(str, len);
				int head = br.ReadInt();
				int groupID = br.ReadShort();
				int serviceID = br.ReadShort();
				BaseService service = null;
				if (m_services.containsKey(serviceID)) {
					m_services.get(serviceID).OnCallBack(br, socketID, localSID, len);
				}
				br.Close();
			}
		} catch (Exception ex) {
			System.out.println(ex.getMessage() + "\r\n" + ex.getStackTrace());
		}
	}

	/**
	 * 销毁对象
	 * 
	 */
	public void Dispose() {
		if (!m_isDisposed) {
			m_listeners.clear();
			m_isDisposed = true;
		}
	}

	/**
	 * 获取网页数据
	 * 
	 * @param url
	 *            地址
	 * @return 页面源码
	 */
	public static String GetHttpWebRequest(String url) {
		return "";
	}

	/**
	 * 获取请求ID
	 * 
	 * @return 请求ID
	 */
	public static int GetRequestID() {
		return m_requestID++;
	}

	/**
	 * 获取所有的服务
	 * 
	 * @param services
	 *            服务列表
	 */
	public static void GetServices(java.util.ArrayList<BaseService> services) {
		for (BaseService service : m_services.values()) {
			services.add(service);
		}
	}

	/**
	 * 保持活跃
	 * 
	 * @param socketID
	 *            套接字ID
	 * @return 状态
	 */
	public int KeepAlive(int socketID) {
		int ret = -1;
		try {
			Binary bw = new Binary();
			bw.WriteInt((int) 4);
			ret = SendByServer(socketID, m_socketID, bw.GetBytes(), 4);
			bw.Close();
		} catch (Exception ex) {

		}
		return ret;
	}

	/**
	 * 收到消息
	 * 
	 * @param br
	 *            流
	 * @param socketID
	 *            套接字ID
	 * @param localSID
	 *            本地套接字ID
	 * @param len
	 *            长度
	 */
	public void OnCallBack(Binary br, int socketID, int localSID, int len) {
		try {
			int headSize = 4 * 4 + 2 * 3 + 1 * 2;
			int functionID = br.ReadShort();
			int sessionID = br.ReadInt();
			int requestID = br.ReadInt();
			int state = (int) br.ReadChar();
			int compressType = (int) br.ReadChar();
			int bodyLength = br.ReadInt();
			byte[] body = new byte[len - headSize];
			br.ReadBytes(body);
			CMessage message = null;
			byte[] unzipBody = null;
			if (compressType == COMPRESSTYPE_GZIP) {
				unzipBody = CStrA.UnGZip(body);
				bodyLength = unzipBody.length;
				message = new CMessage(GetGroupID(), GetServiceID(), functionID, sessionID, requestID, socketID, state,
						compressType, bodyLength, unzipBody);
				// TODO...
			} else {
				message = new CMessage(GetGroupID(), GetServiceID(), functionID, sessionID, requestID, socketID, state,
						compressType, bodyLength, body);
			}

			OnReceive(message);
			OnWaitMessageHandle(message);
			body = null;
			unzipBody = null;
		} catch (Exception ex) {

		}
	}
	
	/**
	 * 客户端退出方法
	 * 
	 */
	public void OnClientClose(int socketID, int localSID)
	{
        if (m_compressTypes.containsKey(socketID))
        {
            m_compressTypes.remove(socketID);
        }
	}

	/**
	 * 接收数据
	 * 
	 * @param message
	 *            消息
	 */
	public void OnReceive(CMessage message) {
		synchronized (m_compressTypes)
        {
            m_compressTypes.put(message.m_socketID, message.m_compressType);
        }
	}

	/**
	 * 等待消息的处理
	 * 
	 * @param message
	 *            消息
	 */
	public void OnWaitMessageHandle(CMessage message) {
		if (m_waitMessages.size() > 0) {
			synchronized (m_waitMessages) {
				if (m_waitMessages.containsKey(message.m_requestID)) {
					CMessage waitMessage = m_waitMessages.get(message.m_requestID);
					waitMessage.Copy(message);
				}
			}
		}
	}

	/**
	 * 注册数据监听
	 * 
	 * @param requestID
	 *            请求ID
	 * @param callBack
	 *            回调函数
	 */
	public void RegisterListener(int requestID, ListenerMessageCallBack callBack) {
		synchronized (m_listeners) {
			MessageListener listener = null;
			if (!m_listeners.containsKey(requestID)) {
				listener = new MessageListener();
				m_listeners.put(requestID, listener);
			} else {
				listener = m_listeners.get(requestID);
			}
			listener.Add(callBack);
		}
	}

	/**
	 * 注册等待
	 * 
	 * @param requestID
	 *            请求ID
	 * @param message
	 *            消息
	 */
	public void RegisterWait(int requestID, CMessage message) {
		synchronized (m_waitMessages) {
			m_waitMessages.put(requestID, message);
		}
	}

	/**
	 * 发送消息
	 * 
	 * @param message
	 *            消息
	 */
	public int Send(CMessage message) {
		int ret = -1;
		try {
			Binary bw = new Binary();
			byte[] body = message.m_body;
			int bodyLength = message.m_bodyLength;
			int uncBodyLength = bodyLength;
			synchronized (m_compressTypes)
            {
                if (m_compressTypes.containsKey(message.m_socketID))
                {
                    message.m_compressType = m_compressTypes.get(message.m_socketID);
                }
            }
			byte[] zipBody = null;
			if (message.m_compressType == COMPRESSTYPE_GZIP) {
				zipBody = CStrA.GZip(body);
				bodyLength = zipBody.length;
			}
			int len = 4 * 4 + bodyLength + 2 * 3 + 1 * 2;
			bw.WriteInt(len);
			bw.WriteShort((short) message.m_groupID);
			bw.WriteShort((short) message.m_serviceID);
			bw.WriteShort((short) message.m_functionID);
			bw.WriteInt(message.m_sessionID);
			bw.WriteInt(message.m_requestID);
			bw.WriteChar((char) message.m_state);
			bw.WriteChar((char) message.m_compressType);
			bw.WriteInt(uncBodyLength);
			if (message.m_compressType == COMPRESSTYPE_GZIP) {
				bw.WriteBytes(zipBody);
			} else {
				bw.WriteBytes(body);
			}
			byte[] bytes = bw.GetBytes();
			ret = SendByServer(message.m_socketID, m_socketID, bytes, bytes.length);
			bw.Close();
		} catch (Exception ex) {

		}
		return ret;
	}

	/**
	 * 发送到监听者
	 * 
	 * @param message
	 *            消息
	 */
	public void SendToListener(CMessage message) {
		MessageListener listener = null;
		synchronized (m_listeners) {
			if (m_listeners.containsKey(message.m_requestID)) {
				listener = m_listeners.get(message.m_requestID);
			}
		}
		if (listener != null) {
			listener.CallBack(message);
		}
	}

	/**
	 * 取消注册数据监听
	 * 
	 * @param requestID
	 *            请求ID
	 */
	public void UnRegisterListener(int requestID) {
		synchronized (m_listeners) {
			m_listeners.remove(requestID);
		}
	}

	/**
	 * 取消注册监听
	 * 
	 * @param requestID
	 *            请求ID
	 * @param callBack
	 *            回调函数
	 */
	public void UnRegisterListener(int requestID, ListenerMessageCallBack callBack) {
		synchronized (m_listeners) {
			if (m_listeners.containsKey(requestID)) {
				m_listeners.get(requestID).Remove(callBack);
			}
		}
	}

	/**
	 * 取消注册等待
	 * 
	 * @param requestID
	 *            请求ID
	 */
	public void UnRegisterWait(int requestID) {
		synchronized (m_waitMessages) {
			if (m_waitMessages.containsKey(requestID)) {
				m_waitMessages.remove(requestID);
			}
		}
	}

	/**
	 * 等待消息
	 * 
	 * @param requestID
	 *            请求ID
	 * @param timeout
	 *            超时
	 * @return 状态
	 */
	public int WaitMessage(int requestID, int timeout) {
		int state = 0;
		try {
			while (timeout > 0) {
				synchronized (m_waitMessages) {
					if (m_waitMessages.containsKey(requestID)) {
						if (m_waitMessages.get(requestID).m_bodyLength > 0) {
							state = 1;
							break;
						}
					} else {
						break;
					}
				}
				timeout -= 10;
				Thread.sleep(10);
			}
			UnRegisterWait(requestID);
		} catch (Exception ex) {
		}
		return state;
	}

	/**
	 * 等待消息
	 * 
	 * @param socketID
	 *            目标SocketID
	 * @param localSID
	 *            本地SocketID
	 * @param state
	 *            状态
	 * @param log
	 *            日志
	 * @return 状态
	 */
	public static void WriteLog(int socketID, int localSID, int state, String log) {
		if(state == 2)
		{
			for (BaseService service : m_services.values()) {
				service.OnClientClose(socketID, localSID);
			}
		}
	}
}